File: apps/agent/src/agent/react-loop.ts
import { randomUUID } from "node:crypto";
import { streamGenerate, type Content, type ContentPart } from "../lib/vertex.js";
import { SYSTEM_PROMPT, userFraming } from "./prompts.js";
import { TOOL_REGISTRY, TOOLS } from "./tools/index.js";
import type { AgentEvent, Citation } from "./types.js";

const MAX_TURNS = 14;

export interface RunOptions {
  query: string;
  maxTurns?: number;
  /** Override the system prompt — used by the /debate route to spawn a
      Devil's Advocate variant on the same query. Defaults to SYSTEM_PROMPT. */
  systemPrompt?: string;
  /** Prior conversation turns. Passed verbatim as Gemini `contents` before
      the current user query. Each entry is what the user asked + what the
      agent answered. Used by /ask for multi-turn follow-ups. */
  history?: Array<{ role: "user" | "model"; text: string }>;
}

export async function* runAgent(
  opts: RunOptions,
): AsyncGenerator<AgentEvent, void, unknown> {
  const startedAt = Date.now();
  const runId = randomUUID();
  const maxTurns = opts.maxTurns ?? MAX_TURNS;
  const declarations = TOOLS.map((t) => t.declaration);
  const contents: Content[] = [];
  // Thread prior conversation turns first (if any). Each historical turn
  // is a user message followed by the model's answer — Gemini handles the
  // rest as native multi-turn context.
  if (opts.history && opts.history.length > 0) {
    for (const turn of opts.history) {
      contents.push({ role: turn.role, parts: [{ text: turn.text }] });
    }
  }
  contents.push({ role: "user", parts: [{ text: userFraming(opts.query) }] });
  const allCitations: Citation[] = [];
  const usage = { prompt: 0, candidates: 0, thoughts: 0, total: 0 };
  // Every draft_email must be audited by the Critic. We track which drafted
  // actions have already been critiqued (by the model OR by us) so the
  // invariant is enforced in code, not just requested in the prompt.
  const critiquedActionIds = new Set<string>();

  yield { kind: "start", query: opts.query, runId, at: Date.now() };

  let turn = 0;
  while (turn < maxTurns) {
    turn++;
    const collected: ContentPart[] = [];
    let pendingText = "";
    const toolCalls: Array<{ name: string; args: Record<string, unknown> }> = [];

    try {
      for await (const chunk of streamGenerate({
        system: opts.systemPrompt ?? SYSTEM_PROMPT,
        contents,
        tools: declarations,
        temperature: 0.4,
        maxTokens: 2048,
      })) {
        if (chunk.text !== undefined && chunk.text.length > 0) {
          yield {
            kind: "thought",
            chunk: chunk.text,
            at: Date.now(),
          };
          if (chunk.thoughtSignature) {
            // A signed text part: flush any pending unsigned text, then push
            // this part with its signature preserved (required by Gemini 3.x).
            if (pendingText.length > 0) {
              collected.push({ text: pendingText });
              pendingText = "";
            }
            collected.push({
              text: chunk.text,
              thoughtSignature: chunk.thoughtSignature,
            });
          } else {
            pendingText += chunk.text;
          }
        }
        if (chunk.functionCall) {
          if (pendingText.length > 0) {
            collected.push({ text: pendingText });
            pendingText = "";
          }
          toolCalls.push(chunk.functionCall);
          collected.push({
            functionCall: chunk.functionCall,
            ...(chunk.thoughtSignature ? { thoughtSignature: chunk.thoughtSignature } : {}),
          });
        }
        if (chunk.usage) {
          // Vertex sometimes emits cumulative usage in later chunks of the same
          // turn — overwrite per-turn then add at the end. Here we track running
          // max for this turn and roll into the run total below.
          if (chunk.usage.promptTokens !== undefined) usage.prompt += chunk.usage.promptTokens;
          if (chunk.usage.candidatesTokens !== undefined) usage.candidates += chunk.usage.candidatesTokens;
          if (chunk.usage.thoughtsTokens !== undefined) usage.thoughts += chunk.usage.thoughtsTokens;
          if (chunk.usage.totalTokens !== undefined) usage.total += chunk.usage.totalTokens;
        }
      }
    } catch (err) {
      const msg = err instanceof Error ? err.message : String(err);
      yield { kind: "error", message: msg, at: Date.now() };
      return;
    }

    if (pendingText.length > 0) {
      collected.push({ text: pendingText });
    }

    // No tool calls → this is the final answer turn.
    if (toolCalls.length === 0) {
      const finalText = collected
        .map((p) => p.text ?? "")
        .filter((t) => t.length > 0)
        .join("");
      if (finalText.length === 0) {
        yield {
          kind: "error",
          message: "model returned no content and no tool call",
          at: Date.now(),
        };
        return;
      }
      if (allCitations.length > 0) {
        yield { kind: "citations", citations: dedupCitations(allCitations), at: Date.now() };
      }
      yield { kind: "answer", chunk: finalText, at: Date.now() };
      yield {
        kind: "done",
        turns: turn,
        totalMs: Date.now() - startedAt,
        usage: {
          promptTokens: usage.prompt,
          candidatesTokens: usage.candidates,
          thoughtsTokens: usage.thoughts,
          totalTokens: usage.total > 0 ? usage.total : usage.prompt + usage.candidates + usage.thoughts,
          estimatedCostUsd: estimateCost(usage.prompt, usage.candidates + usage.thoughts),
        },
        at: Date.now(),
      };
      return;
    }

    // Record the full model turn (text + all function calls) verbatim.
    contents.push({ role: "model", parts: collected });

    // Execute every tool call in order, collecting one functionResponse per call.
    // Gemini 3.x requires the next user turn to contain a functionResponse for
    // *every* functionCall in the model turn — same count, same order.
    const responseParts: ContentPart[] = [];
    // Drafts produced this turn that still need a Critic pass.
    const draftedThisTurn: string[] = [];
    for (const call of toolCalls) {
      const callId = randomUUID().slice(0, 8);
      const tool = TOOL_REGISTRY.get(call.name);

      yield {
        kind: "tool_call",
        id: callId,
        name: call.name,
        args: call.args,
        at: Date.now(),
      };

      const t0 = Date.now();
      let result;
      if (!tool) {
        result = { ok: false, error: `unknown tool: ${call.name}` };
      } else {
        try {
          result = await tool.handler(call.args, { query: opts.query, runId });
        } catch (err) {
          const msg = err instanceof Error ? err.message : String(err);
          result = { ok: false, error: msg };
        }
      }
      const durationMs = Date.now() - t0;

      // Track the draft/critique pairing so we can enforce the audit below.
      if (call.name === "critique_draft") {
        const aid = typeof call.args["action_id"] === "string" ? (call.args["action_id"] as string) : "";
        if (aid) critiquedActionIds.add(aid);
      } else if (call.name === "draft_email" && result.ok) {
        const aid = typeof result.data?.["actionId"] === "string" ? (result.data["actionId"] as string) : "";
        if (aid) draftedThisTurn.push(aid);
      }

      if (result.citations && result.citations.length > 0) {
        allCitations.push(...result.citations);
      }

      yield {
        kind: "observation",
        id: callId,
        name: call.name,
        result,
        durationMs,
        at: Date.now(),
      };

      responseParts.push({
        functionResponse: {
          name: call.name,
          response: trimForModel(result),
        },
      });
    }

    // ── ENFORCE THE CRITIC ──
    // For every draft_email this turn that the model did NOT pair with its own
    // critique_draft call, run the Critic ourselves and feed the verdict back
    // into the same user turn. This guarantees no draft ever reaches the user
    // un-audited, regardless of whether the model followed the prompt.
    const critic = TOOL_REGISTRY.get("critique_draft");
    for (const actionId of draftedThisTurn) {
      if (critiquedActionIds.has(actionId) || !critic) continue;
      const callId = randomUUID().slice(0, 8);
      // `auto: true` lets the UI distinguish an enforced critique from one the
      // model requested itself.
      yield {
        kind: "tool_call",
        id: callId,
        name: "critique_draft",
        args: { action_id: actionId, auto: true },
        at: Date.now(),
      };
      const t0 = Date.now();
      let result;
      try {
        result = await critic.handler({ action_id: actionId }, { query: opts.query, runId });
      } catch (err) {
        const msg = err instanceof Error ? err.message : String(err);
        result = { ok: false, error: msg };
      }
      const durationMs = Date.now() - t0;
      critiquedActionIds.add(actionId);

      if (result.citations && result.citations.length > 0) {
        allCitations.push(...result.citations);
      }

      yield {
        kind: "observation",
        id: callId,
        name: "critique_draft",
        result,
        durationMs,
        at: Date.now(),
      };

      responseParts.push({
        functionResponse: {
          name: "critique_draft",
          response: trimForModel(result),
        },
      });
    }

    contents.push({ role: "user", parts: responseParts });
  }

  yield {
    kind: "error",
    message: `exceeded max turns (${maxTurns}) without final answer`,
    at: Date.now(),
  };
}

function trimForModel(result: {
  ok: boolean;
  data?: Record<string, unknown>;
  error?: string;
  summary?: string;
}): Record<string, unknown> {
  if (!result.ok) {
    return { ok: false, error: result.error ?? "tool failed" };
  }
  return {
    ok: true,
    ...(result.summary ? { summary: result.summary } : {}),
    ...(result.data ? { data: result.data } : {}),
  };
}

function dedupCitations(list: Citation[]): Citation[] {
  const seen = new Set<string>();
  const out: Citation[] = [];
  for (const c of list) {
    if (!c.chunkId || seen.has(c.chunkId)) continue;
    seen.add(c.chunkId);
    out.push(c);
  }
  return out.sort((a, b) => b.score - a.score).slice(0, 12);
}

/**
 * Cost estimate for Gemini 3 Pro Preview (per Vertex pricing as of May 2026):
 *   input  $1.25 / 1M tokens   (≤ 200k input)
 *   output $10.00 / 1M tokens  (≤ 200k input)
 * Thinking tokens are billed as output. We round the result so the chip
 * displays as $0.0021 not $0.00214837.
 */
function estimateCost(promptTokens: number, outputTokens: number): number {
  const inUsd = (promptTokens / 1_000_000) * 1.25;
  const outUsd = (outputTokens / 1_000_000) * 10.0;
  const total = inUsd + outUsd;
  return Math.round(total * 10000) / 10000;
}

File: apps/agent-py/app/agent/react_loop.py
"""ReAct loop — port of apps/agent/src/agent/react-loop.ts.

Hand-rolled think→act→observe over google-genai streaming + function calling.
Yields AgentEvent dicts the SSE route serializes. Enforces the Critic: every
draft_email the model doesn't pair with its own critique_draft is critiqued
automatically and fed back into the same turn.
"""
from __future__ import annotations

import json
import time
import uuid
from typing import AsyncIterator

from app.agent.prompts import SYSTEM_PROMPT, user_framing
from app.agent.tools.registry import DECLARATIONS, TOOL_REGISTRY
from app.llm.genai_client import stream_generate

MAX_TURNS = 14


def _now() -> int:
    return int(time.time() * 1000)


def _estimate_cost(prompt_tokens: int, output_tokens: int) -> float:
    """Rough USD estimate from public list prices per 1M tokens (input, output),
    keyed by the active generation provider (and model, on Bedrock)."""
    from app.config import llm_provider, settings
    provider = llm_provider()
    if provider == "bedrock":
        mid = settings.bedrock_model_id.lower()
        bedrock_prices = {
            "nova-micro": (0.035, 0.14), "nova-lite": (0.06, 0.24),
            "nova-pro": (0.80, 3.20), "nova-premier": (2.50, 12.50),
            "claude": (3.0, 15.0),
        }
        pin, pout = next((v for k, v in bedrock_prices.items() if k in mid), (0.80, 3.20))
    else:
        pin, pout = {"gemini_api": (0.075, 0.30), "vertex": (1.25, 10.0)}.get(provider, (0.80, 3.20))
    usd = (prompt_tokens / 1_000_000) * pin + (output_tokens / 1_000_000) * pout
    return round(usd * 10000) / 10000


def _trim_for_model(result: dict) -> dict:
    if not result.get("ok"):
        return {"ok": False, "error": result.get("error", "tool failed")}
    out: dict = {"ok": True}
    if result.get("summary"):
        out["summary"] = result["summary"]
    if result.get("data"):
        out["data"] = result["data"]
    return out


def _dedup_citations(cites: list[dict]) -> list[dict]:
    seen: set[str] = set()
    out: list[dict] = []
    for c in cites:
        cid = c.get("chunkId")
        if not cid or cid in seen:
            continue
        seen.add(cid)
        out.append(c)
    out.sort(key=lambda c: c.get("score", 0), reverse=True)
    return out[:12]


async def run_agent(
    query: str,
    *,
    max_turns: int = MAX_TURNS,
    system_prompt: str | None = None,
    history: list[dict] | None = None,
) -> AsyncIterator[dict]:
    started = time.time()
    run_id = str(uuid.uuid4())
    # Provider-neutral conversation (app.llm.neutral): each backend converts at
    # call time, so the loop is identical on Bedrock, Gemini, and Vertex.
    messages: list[dict] = []
    if history:
        for turn in history:
            role = "assistant" if turn["role"] in ("assistant", "model") else "user"
            messages.append({"role": role, "parts": [{"text": turn["text"]}]})
    messages.append({"role": "user", "parts": [{"text": user_framing(query)}]})

    all_citations: list[dict] = []
    usage = {"prompt": 0, "candidates": 0, "thoughts": 0, "total": 0}
    critiqued_action_ids: set[str] = set()

    yield {"kind": "start", "query": query, "runId": run_id, "at": _now()}

    turn = 0
    while turn < max_turns:
        turn += 1
        collected_text: list[str] = []
        tool_calls: list[dict] = []  # each {"id", "name", "args"}

        try:
            async for chunk in stream_generate(
                system=system_prompt or SYSTEM_PROMPT,
                contents=messages,
                tools=DECLARATIONS,
                temperature=0.4,
                max_tokens=2048,
            ):
                if chunk.text:
                    yield {"kind": "thought", "chunk": chunk.text, "at": _now()}
                    collected_text.append(chunk.text)
                if chunk.function_call:
                    tool_calls.append(chunk.function_call)
                if chunk.usage:
                    usage.update({
                        "prompt": chunk.usage["prompt"], "candidates": chunk.usage["candidates"],
                        "thoughts": chunk.usage["thoughts"], "total": chunk.usage["total"],
                    })
        except Exception as err:  # noqa: BLE001
            yield {"kind": "error", "message": str(err), "at": _now()}
            return

        # No tool calls → final answer turn.
        if not tool_calls:
            final_text = "".join(collected_text)
            if not final_text:
                yield {"kind": "error", "message": "model returned no content and no tool call", "at": _now()}
                return
            if all_citations:
                yield {"kind": "citations", "citations": _dedup_citations(all_citations), "at": _now()}
            yield {"kind": "answer", "chunk": final_text, "at": _now()}
            out_tokens = usage["candidates"] + usage["thoughts"]
            yield {"kind": "done", "turns": turn, "totalMs": int((time.time() - started) * 1000),
                   "usage": {
                       "promptTokens": usage["prompt"], "candidatesTokens": usage["candidates"],
                       "thoughtsTokens": usage["thoughts"],
                       "totalTokens": usage["total"] or (usage["prompt"] + out_tokens),
                       "estimatedCostUsd": _estimate_cost(usage["prompt"], out_tokens),
                   }, "at": _now()}
            return

        # Record the model turn (text + tool calls) as a neutral assistant message.
        assistant_parts: list = []
        joined = "".join(collected_text)
        if joined:
            assistant_parts.append({"text": joined})
        for tc in tool_calls:
            assistant_parts.append({"tool_call": tc})
        messages.append({"role": "assistant", "parts": assistant_parts})

        response_parts: list = []
        drafted_this_turn: list[str] = []
        for call in tool_calls:
            call_id = call["id"]  # matches the tool_call id in the assistant turn
            tool = TOOL_REGISTRY.get(call["name"])
            yield {"kind": "tool_call", "id": call_id, "name": call["name"], "args": call["args"], "at": _now()}

            t0 = time.time()
            if not tool:
                result = {"ok": False, "error": f"unknown tool: {call['name']}"}
            else:
                try:
                    result = await tool.handler(call["args"], {"query": query, "runId": run_id})
                except Exception as err:  # noqa: BLE001
                    result = {"ok": False, "error": str(err)}
            duration_ms = int((time.time() - t0) * 1000)

            if call["name"] == "critique_draft":
                aid = call["args"].get("action_id")
                if isinstance(aid, str) and aid:
                    critiqued_action_ids.add(aid)
            elif call["name"] == "draft_email" and result.get("ok"):
                aid = (result.get("data") or {}).get("actionId")
                if isinstance(aid, str) and aid:
                    drafted_this_turn.append(aid)

            if result.get("citations"):
                all_citations.extend(result["citations"])

            yield {"kind": "observation", "id": call_id, "name": call["name"],
                   "result": result, "durationMs": duration_ms, "at": _now()}
            response_parts.append({"tool_result": {
                "id": call_id, "name": call["name"], "content": _trim_for_model(result)}})

        # ── ENFORCE THE CRITIC ──
        # The model didn't call critique_draft itself, so there's no matching
        # tool_call — feed the audit back as a text note (a synthetic tool_result
        # would be rejected by Bedrock's strict tool protocol).
        critic = TOOL_REGISTRY.get("critique_draft")
        for action_id in drafted_this_turn:
            if action_id in critiqued_action_ids or not critic:
                continue
            call_id = uuid.uuid4().hex[:8]
            yield {"kind": "tool_call", "id": call_id, "name": "critique_draft",
                   "args": {"action_id": action_id, "auto": True}, "at": _now()}
            t0 = time.time()
            try:
                result = await critic.handler({"action_id": action_id}, {"query": query, "runId": run_id})
            except Exception as err:  # noqa: BLE001
                result = {"ok": False, "error": str(err)}
            duration_ms = int((time.time() - t0) * 1000)
            critiqued_action_ids.add(action_id)
            if result.get("citations"):
                all_citations.extend(result["citations"])
            yield {"kind": "observation", "id": call_id, "name": "critique_draft",
                   "result": result, "durationMs": duration_ms, "at": _now()}
            response_parts.append({"text":
                f"[auto-critic] Audit of draft {action_id}: {json.dumps(_trim_for_model(result))}"})

        messages.append({"role": "user", "parts": response_parts})

    yield {"kind": "error", "message": f"exceeded max turns ({max_turns}) without final answer", "at": _now()}
